recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,557 @@
+                output_beam=self.hparams.test_output_beam,
+                ac_scale=self.hparams.ac_scale,
+                max_active_states=self.hparams.test_max_active_state,
+                min_active_states=self.hparams.test_min_active_state,
+            )
+            paths = self.decoder["decoding_method"](lattice)
+
+        if stage == sb.Stage.TEST or stage == sb.Stage.VALID:
+            for k, path in paths.items():
+                predicted_texts = sbk2.utils.lattice_paths_to_text(
+                    path, self.lexicon.word_table
+                )
+
+                predicted_words = [wrd.split(" ") for wrd in predicted_texts]
+                target_words = [wrd.split(" ") for wrd in texts]
+                self.wer_metrics[k].append(ids, predicted_words, target_words)
+                self.cer_metrics[k].append(ids, predicted_words, target_words)
+            # For TEST and VALID stages, the loss value is not exact.
+            # The <UNK> words have a target length (e.g., number of phones or characters) of 1.
+            # As such, sentences with <UNK> have a higher loss during CTC loss 'mean' reduction mode.
+            # It does not impact training.
+        return loss
+
+    def fit_batch(self, batch):
+        should_step = self.step % self.grad_accumulation_factor == 0
+
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            self.wav2vec_optimizer.zero_grad()
+            self.model_optimizer.zero_grad()
+            with torch.cuda.amp.autocast():
+                with self.no_sync():
+                    outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+                loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            with self.no_sync(not should_step):
+                self.scaler.scale(
+                    loss / self.grad_accumulation_factor
+                ).backward()
+            if should_step:
+                if not self.hparams.freeze_wav2vec:
+                    self.scaler.unscale_(self.wav2vec_optimizer)
+                self.scaler.unscale_(self.model_optimizer)
+                if self.check_gradients(loss):
+                    self.scaler.step(self.wav2vec_optimizer)
+                    self.scaler.step(self.model_optimizer)
+                self.scaler.update()
+                self.optimizer_step += 1
+        else:
+            with self.no_sync():
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+            (loss / self.grad_accumulation_factor).backward()
+            if should_step:
+                if self.check_gradients(loss):
+                    self.wav2vec_optimizer.step()
+                    self.model_optimizer.step()
+                self.wav2vec_optimizer.zero_grad()
+                self.model_optimizer.zero_grad()
+                self.optimizer_step += 1
+
+        return loss.detach().cpu()
```

Comment:
this can me removed, please see `train_with_wav2vec.py` (unstable-v0.6 branch)

Reply:
DONE